tests/src/app/testqgsmaptooleditmesh.cpp

```diff
@@ -0,0 +1,288 @@
+/***************************************************************************
+  testqgsmaptooleditmesh.cpp - %{Cpp:License:ClassName}
+
+ ---------------------
+ begin                : 25.6.2021
+ copyright            : (C) 2021 by Vincent Cloarec
+ email                : vcloarec at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstest.h"
+#include "qgisapp.h"
+#include "testqgsmaptoolutils.h"
+#include "qgsmaptooleditmeshframe.h"
+#include "qgsmeshlayer.h"
+#include "qgsmesheditor.h"
+
+class TestQgsMapToolEditMesh : public QObject
+{
+    Q_OBJECT
+  public:
+    TestQgsMapToolEditMesh() = default;
+
+  private slots:
+    void initTestCase();// will be called before the first testfunction is executed.
+    void cleanupTestCase() {};// will be called after the last testfunction was executed.
+    void init(); // will be called before each testfunction is executed.
+    void cleanup() {} // will be called after every testfunction.
+
+    void editMesh();
+
+  private:
+    QgisApp *mQgisApp = nullptr;
+    std::unique_ptr<QgsMeshLayer> meshLayerQuadFlower;
+    QString mDataDir;
+    std::unique_ptr<QgsMapCanvas> canvas;
+    QgsMapToolEditMeshFrame *editMeshMapTool;
+};
+
+
+//runs before all tests
+void TestQgsMapToolEditMesh::initTestCase()
+{
+  QgsApplication::init();
+  QgsApplication::initQgis();
+  mQgisApp = new QgisApp();
+  mDataDir = QString( TEST_DATA_DIR ); //defined in CmakeLists.txt
+  mDataDir += "/mesh";
+  canvas.reset( new QgsMapCanvas() );
+  editMeshMapTool = new QgsMapToolEditMeshFrame( canvas.get() );
+}
+
+void TestQgsMapToolEditMesh::init()
+{
+  QString uri = QString( mDataDir + "/quad_flower.2dm" );
+  meshLayerQuadFlower.reset( new QgsMeshLayer( uri, "Quad Flower", "mdal" ) );
+  QVERIFY( meshLayerQuadFlower );
+  QCOMPARE( meshLayerQuadFlower->datasetGroupCount(), 1 );
+
+  QgsCoordinateTransform transform;
+  meshLayerQuadFlower->startFrameEditing( transform );
+
+  canvas->setLayers( QList<QgsMapLayer *>() << meshLayerQuadFlower.get() );
+}
+
+void TestQgsMapToolEditMesh::editMesh()
```

Comment:
can we split it to the separate tests for each action?

Reply:
actions could depend of the previous one